netperf/man1/netperf.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log$
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database netperf.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so netperf.macros
.\"
.\"
.TH NETPERF 1L "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ User Commands"
.\" =========================================================================
.\" =========================================================================
.SH NAME
.B netperf
\- a network performance benchmark
.\" =========================================================================
.\" =========================================================================
.SH SYNOPSIS
.PP
.HP 12
\fBnetperf\fR [\fB-t\fR\ \fITESTNAME\fR] [\fIglobal-test-options\fR] [-- [\fItest-specific-options\fR]]
.HP 12
\fBnetperf\fR \fB-h\fR,\ \fB--help\fR
.HP 12
\fBnetperf\fR [\fB-t\fR\ \fITESTNAME\fR]
[\fB-a\fR\ [\fILOC-SND-ALIGN\fR][\fB,\fR][\fILOC-RCV-ALIGN\fR]]
[\fB-A\fR\ [\fIREM-SND-ALIGN\fR][\fB,\fR][\fIREM-RCV-ALIGN\fR]]
[\fB-c\fR\ [\fILOC-RATE\fR]]
[\fB-C\fR\ [\fIREM-RATE\fR]]
[\fB-d\fR]
[\fB-f\fR\ {\ \fBG\fR\ |\ \fBM\fR\ |\ \fBK\fR\ |\ \fBg\fR\ |\ \fBm\fR\ |\ \fBk\fR\ }]
[\fB-F\fR\ \fIFILL-FILE\fR]
[\fB-H\fR\ \fIREMOTE-HOST\fR]
[\fB-i\fR\ \fIMAX\fB,\fIMIN\fR]
[\fB-I\fR\ \fILEVEL\fR[\fB,\fIWIDTH\fR]]
[\fB-l\fR\ \fITESTLEN\fR]
[\fB-n\fR\ \fINUM-CPUS\fR]
[\fB-o\fR\ [\fILOC-SND-OFFSET\fR][\fB,\fR][\fILOC-RCV-OFFSET\fR]]
[\fB-O\fR\ [\fIREM-SND-OFFSET\fR][\fB,\fR][\fIREM-RCV-OFFSET\fR]]
[\fB-p\fR\ \fIPORTNUM\fR]
[\fB-P\fR\ {\ \fB0\fR\ |\ \fB1\fR\ }]
[\fB-v\fR\ \fIVERBOSITY\fR]
[--\ [\fItest-specific-options\fR]]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIsocket-test\fR [\fIglobal-test-options\fR]
[--
[\fB-CDVz\fR]
.br
[\fB-b\fR\ \fIBURST-SIZE\fR]
[\fB-I\fR\ [\fILOC-IP-ADDR\fR[:\fILOC-IP-ADDR\fR]*][\fB,\fR[\fIREM-IP-ADDR\fR][:\fIREM-IP-ADDR\fR]*]]
[\fB-m\fR\ \fILOC-MSG-SIZE\fR]
[\fB-M\fR\ \fIREM-MSG-SIZE\fR]
[\fB-p\fR\ [\fIMIN-PORTNUM\fR][\fB,\fR][\fIMAX-PORTNUM\fR]]
[\fB-P\fR\ [\fILOC-PORTNUM\fR][\fB,\fR][\fIREM-PORTNUM\fR]]
[\fB-r\fR\ [\fIREQ-SIZE\fR][\fB,\fR][\fIRSP-SIZE\fR]]
[\fB-s\fR\ [\fILOC-SNDBUF\fR][\fB,\fR][\fILOC-RCVBUF\fR]]
[\fB-S\fR\ [\fIREM-SNDBUF\fR][\fB,\fR][\fIREM-RCVBUF\fR]]
[\fB-w\fR\ \fIRCV-WIDTH\fR]
[\fB-W\fR\ \fISND-WIDTH\fR]
]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIxti-test\fR [\fIglobal-test-options\fR]
[--
[\fB-DV\fR]
.br
[\fB-I\fR\ [\fILOC-IP-ADDR\fR[:\fILOC-IP-ADDR\fR]*][\fB,\fR[\fIREM-IP-ADDR\fR][:\fIREM-IP-ADDR\fR]*]]
[\fB-m\fR\ \fILOC-MSG-SIZE\fR]
[\fB-M\fR\ \fIREM-MSG-SIZE\fR]
[\fB-r\fR\ [\fIREQ-SIZE\fR][\fB,\fR][\fIRSP-SIZE\fR]]
[\fB-s\fR\ [\fILOC-SNDBUF\fR][\fB,\fR][\fILOC-RCVBUF\fR]]
[\fB-S\fR\ [\fIREM-SNDBUF\fR][\fB,\fR][\fIREM-RCVBUF\fR]]
[\fB-w\fR\ \fIRCV-WIDTH\fR]
[\fB-W\fR\ \fISND-WIDTH\fR]
[\fB-X\fR\ [\fILOC-DEVICE\fR][\fB,\fR][\fIREM-DEVICE\fR]]
]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIipv6-test\fR [\fIglobal-test-options\fR]
[--
[\fB-DV\fR]
.br
[\fB-H\fR\ \fIHOSTNAME\fR]
[\fB-m\fR\ \fILOC-MSG-SIZE\fR]
[\fB-M\fR\ \fIREM-MSG-SIZE\fR]
[\fB-p\fR\ [\fIMIN-PORTNUM\fR][\fB,\fR][\fIMAX-PORTNUM\fR]]
[\fB-r\fR\ [\fIREQ-SIZE\fR][\fB,\fR][\fIRSP-SIZE\fR]]
[\fB-s\fR\ [\fILOC-SNDBUF\fR][\fB,\fR][\fILOC-RCVBUF\fR]]
[\fB-S\fR\ [\fIREM-SNDBUF\fR][\fB,\fR][\fIREM-RCVBUF\fR]]
[\fB-w\fR\ \fIRCV-WIDTH\fR]
[\fB-W\fR\ \fISND-WIDTH\fR]
]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIunix-test\fR [\fIglobal-test-options\fR]
[--
.br
[\fB-m\fR\ \fILOC-MSG-SIZE\fR]
[\fB-M\fR\ \fIREM-MSG-SIZE\fR]
[\fB-p\fR\ \fIPATH\fR]
[\fB-s\fR\ [\fILOC-SNDBUF\fR][\fB,\fR][\fILOC-RCVBUF\fR]]
[\fB-S\fR\ [\fIREM-SNDBUF\fR][\fB,\fR][\fIREM-RCVBUF\fR]]
[\fB-r\fR\ [\fIREQ-SIZE\fR][\fB,\fR][\fIRSP-SIZE\fR]]
]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIdlpi-test\fR [\fIglobal-test-options\fR]
[--
.br
[\fB-D\fR\ [\fILOC-DEVICE\fR][\fB,\fR][\fIREM-DEVICE\fR]]
[\fB-m\fR\ \fILOC-MSG-SIZE\fR]
[\fB-M\fR\ \fIREM-MSG-SIZE\fR]
[\fB-p\fR\ [\fILOC-PPA\fR][\fB,\fR][\fIREM-PPA\fR]]
[\fB-r\fR\ [\fIREQ-SIZE\fR][\fB,\fR][\fIRSP-SIZE\fR]]
[\fB-s\fR\ \fISAP\fR]
[\fB-w\fR\ [\fILOC-SND-WINDOW\fR][\fB,\fR][\fILOC-RCV-WINDOW\fR]]
[\fB-W\fR\ [\fIREM-SND-WINDOW\fR][\fB,\fR][\fIREM-RCV-WINDOW\fR]]
]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIdns-test\fR [\fIglobal-test-options\fR]
[--
.br
[\fB-f\fR\ \fIFILENAME\fR]
[\fB-H\fR\ \fIHOSTNAME\fR]
]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIatm-test\fR [\fIglobal-test-options\fR]
[--
.br
[\fIatm-options\fR]]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIhippi-test\fR [\fIglobal-test-options\fR]
[--
.br
[\fIhippi-options\fR]]
.HP 12
\fBnetperf\fR \fB-t\fR\ \fIlwp-test\fR [\fIglobal-test-options\fR]
[--
.br
[\fIlwp-options\fR]]
.\" =========================================================================
.\" =========================================================================
.SH DESCRIPTION
.PP
.B netperf
is a benchmark that can be used to measure various aspects of
networking performance.
Currently, its focus is on bulk data transfer and request/response
performance using SCTP, TCP or UDP, and the Berkeley Sockets
interface. In addition, tests for XTI,
.[
xti
.]
DLPI,
.[
dlpi
.]
and Unix Domain
Sockets, tests for IPv6 may be conditionally compiled-in.
.\" =========================================================================
.SS TESTS
.PP
.B netperf
perfoms the following tests:
.PP
.TS
center tab(:);
lblblblb
l|l|l|l.
Test Name:API:Protocol:Test
_
SCTP_STREAM:Sockets:SCTP:STREAM
SCTP_MAERTS:::MAERTS
SCTP_SENDFILE:::SENDFILE
_:::_
SCTP_RR:::RR
SCTP_NBRR:::NBRR
SCTP_CRR:::CRR
SCTP_TRR:::TRR
SCTP_CC:::CC
_::_:_
TCP_STREAM::TCP:STREAM
TCP_MAERTS:::MAERTS
TCP_SENDFILE:::SENDFILE
_:::_
TCP_RR:::RR
TCP_NBRR:::NBRR
TCP_CRR:::CRR
TCP_TRR:::TRR
TCP_CC:::CC
_::_:_
UDP_STREAM::UDP:STREAM
_:::_
UDP_RR:::RR
_:_:_:_
XTI_SCTP_STREAM:XTI:SCTP:STREAM
_:::_
XTI_SCTP_RR:::RR
XTI_SCTP_CRR:::CRR
XTI_SCTP_TRR:::TRR
XTI_SCTP_CC:::CC
_::_:_
XTI_TCP_STREAM::TCP:STREAM
_:::_
XTI_TCP_RR:::RR
XTI_TCP_CRR:::CRR
XTI_TCP_CC:::CC
_::_:_
XTI_UDP_STREAM::UDP:STREAM
_:::_
XTI_UDP_RR:::RR
_:_:_:_
SCTPIPV6_STREAM:IPv6 Sockets:SCTP:STREAM
_:::_
SCTPIPV6_RR:::RR
SCTPIPV6_CRR:::CRR
SCTPIPV6_TRR:::TRR
_::_:_
TCPIPV6_STREAM::TCP:STREAM
_:::_
TCPIPV6_RR:::RR
TCPIPV6_CRR:::CRR
TCPIPV6_TRR:::TRR
_::_:_
UDPIPV6_STREAM::UDP:STREAM
_:::_
UDPIPV6_RR:::RR
_:_:_:_
STREAM_STREAM:Unix Sockets:CO:STREAM
_:::_
STREAM_RR:::RR
_::_:_
DG_STREAM::CL:STREAM
_:::_
DG_RR:::RR
_:_:_:_
DLCO_STREAM:DLPI:CO:STREAM
_:::_
DLCO_RR:::RR
_::_:_
DLCL_STREAM::CL:STREAM
_:::_
DLCL_RR:::RR
_:_:_:_
LOC_CPU:none:none:none
REM_CPU:::
_:_:_:_
.TE
.\" =========================================================================
.SS PROTOCOLS
.PP
.B netperf
tests the following protocols:
.PP
.TS
tab(:);
lbw(12n)lssss.
SCTP:Stream Control Transmission Protocol
TCP:Transmission Control Protocol
UDP:User Datagram Protocol
UNIX:Unix Domain (CO and CL) Protocols
ATM:Asynchronous Transfer Mode
HIPPI:
.TE
.\" =========================================================================
.SS APPLICATION PROGRAMMING INTERFACES
.PP
.B netperf
tests the following APIs:
.PP
.TS
tab(:);
lbw(12n)lssss.
BSD:BSD/POSIX Sockets Interface
XTI:X/Open Transport Interface (XTI/TLI) XNS 5.2 SuSv2
.[
xti
.]
IPV6:POSIX IPv6 Sockets Interface
UNIX:Unix Domain Sockets Interface
DLPI:Data Link Provider Interface
.[
dlpi
.]
ATM:FORE ATM Interface
HiPPI:HP-UX Interface
.TE
.\" =========================================================================
.SS BENCHMARKING PERFORMANCE TESTS
.PP
.B netperf
test the following benchmarks:
.PP
.TS
tab(:);
lbw(12n)lssss.
STREAM:Unidirectional Stream Test (Client to Server)
MAERTS:Unidirectional Stream Test (Server to Client)
SENDFILE:Unidirectional Stream Test (BSD \fBsendfile\fR(2))
RR:Blocking Request/Response

NEWS
OpenSS7 Master Package -- history of user visible changes.  2007-06-24
$Id: NEWS,v 0.9.2.39 2007/06/24 09:21:13 brian Exp $
Copyright (c) 2001-2007  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.

Most of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.  The packages build better
Debian/Ubuntu .deb files.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.

The release provides the following enhancements and fixes:

   - Added the following sub-packages to the master build (and release):

         - `strnsl-0.9.2.3'
         - `strbcm-0.9.2.4'
         - `striso-0.9.2.3'
         - `strsock-0.9.2.3'
         - `strtty-0.9.2.3'
         - `strutil-0.9.2.6'

   - Automated release file generation making for vastly improved and
     timely text documentation present in the release directory.

   - This release candidate includes the changes made to the strsctp
     drivers at the 2006 SCTP Interop at the University of British
     Columbia.  This version was interoperability tested with all
     implementations present.

   - Better support for Ubuntu and recent `gcc' compilers, including
     debian script corrections.

   - Support for most recent 2.6.18 kernels (including Fedora Core 5
     with inode diet patchset).

   - Now builds 32-bit compatibility libraries and tests them against
     64-bit kernel modules and drivers.  The `make installcheck' target
     will now automatically test both 64-bit native and 32-bit
     compatibility versions, one after the other, on 64-bit platforms.

   - Added versions to all library symbols.

   - Many documentation updates for all OpenSS7 packages.

   - Dropped support for `LiS'.


netperf/netperf.c
/*****************************************************************************

 @(#) $RCSfile: netperf.c,v $ $Name:  $($Revision: 1.1.1.8 $) $Date: 2006/05/23 22:43:29 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; version 2 of the License.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2006/05/23 22:43:29 $ by $Author: brian $

 *****************************************************************************/

#ident "@(#) $RCSfile: netperf.c,v $ $Name:  $($Revision: 1.1.1.8 $) $Date: 2006/05/23 22:43:29 $"

static char const ident[] = "$RCSfile: netperf.c,v $ $Name:  $($Revision: 1.1.1.8 $) $Date: 2006/05/23 22:43:29 $";

#ifdef NEED_MAKEFILE_EDIT
#error you must first edit and customize the makefile to your platform
#endif /* NEED_MAKEFILE_EDIT */
/*
 
	   Copyright (C) 1993-2001 Hewlett-Packard Company
                         ALL RIGHTS RESERVED.
 
  The enclosed software and documention includes copyrighted works of
  Hewlett-Packard Co. For as long as you comply with the following
  limitations, you are hereby authorized to (i) use, reproduce, and
  modify the software and documentation, and to (ii) distribute the
  software and documentation, including modifications, for
  non-commercial purposes only.
      
  1.  The enclosed software and documentation is made available at no
      charge in order to advance the general development of
      high-performance networking products.
 
  2.  You may not delete any copyright notices contained in the
      software or documentation. All hard copies, and copies in
      source code or object code form, of the software or
      documentation (including modifications) must contain at least
      one of the copyright notices.
 
  3.  The enclosed software and documentation has not been subjected
      to testing and quality control and is not a Hewlett-Packard Co.
      product. At a future time, Hewlett-Packard Co. may or may not
      offer a version of the software and documentation as a product.
  
  4.  THE SOFTWARE AND DOCUMENTATION IS PROVIDED "AS IS".
      HEWLETT-PACKARD COMPANY DOES NOT WARRANT THAT THE USE,
      REPRODUCTION, MODIFICATION OR DISTRIBUTION OF THE SOFTWARE OR
      DOCUMENTATION WILL NOT INFRINGE A THIRD PARTY'S INTELLECTUAL
      PROPERTY RIGHTS. HP DOES NOT WARRANT THAT THE SOFTWARE OR
      DOCUMENTATION IS ERROR FREE. HP DISCLAIMS ALL WARRANTIES,
      EXPRESS AND IMPLIED, WITH REGARD TO THE SOFTWARE AND THE
      DOCUMENTATION. HP SPECIFICALLY DISCLAIMS ALL WARRANTIES OF
      MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
  
  5.  HEWLETT-PACKARD COMPANY WILL NOT IN ANY EVENT BE LIABLE FOR ANY
      DIRECT, INDIRECT, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES
      (INCLUDING LOST PROFITS) RELATED TO ANY USE, REPRODUCTION,
      MODIFICATION, OR DISTRIBUTION OF THE SOFTWARE OR DOCUMENTATION.
 
*/
char	netperf_id[]="@(#)netperf.c (c) Copyright 1993-2004 Hewlett-Packard Company. Version 2.3";

#include <stdio.h>
#ifdef HAVE_SYS_TYPES_H
# include <sys/types.h>
#endif
#ifdef HAVE_SYS_STAT_H
# include <sys/stat.h>
#endif
#if STDC_HEADERS
# include <stdlib.h>
# include <stddef.h>
#else
# ifdef HAVE_STDLIB_H
#  include <stdlib.h>
# endif
#endif
#ifdef HAVE_STRING_H
# if !STDC_HEADERS && HAVE_MEMORY_H
#  include <memory.h>
# endif
# include <string.h>
#endif
#ifdef HAVE_STRINGS_H
# include <strings.h>
#endif
#ifdef HAVE_INTTYPES_H
# include <inttypes.h>
#else
# ifdef HAVE_STDINT_H
#  include <stdint.h>
# endif
#endif
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#ifdef HAVE_MALLOC_H
# include <malloc.h>
#endif

#ifdef WIN32
#include <winsock2.h>
#include <windows.h>
#endif /* WIN32 */

#include "netsh.h"
#include "netlib.h"
#include "nettest_bsd.h"
#ifdef DO_UNIX
#include "nettest_unix.h"
#endif /* DO_UNIX */
#ifdef DO_FORE
#include "nettest_fore.h"
#endif /* DO_FORE */
#ifdef DO_HIPPI
#include "nettest_hippi.h"
#endif /* DO_HIPPI */
#ifdef DO_XTI
#include "nettest_xti.h"
#endif /* DO_XTI */
#ifdef DO_DLPI
#include "nettest_dlpi.h"
#endif /* DO_DLPI */
#ifdef DO_IPV6
#include "nettest_ipv6.h"
#endif /* DO_IPV6 */
#ifdef DO_DNS
#include "nettest_dns.h"
#endif /* DO_DNS */

 /* this file contains the main for the netperf program. all the other */
 /* routines can be found in the file netsh.c */


int _cdecl
main(int argc, char *argv[])
{

#ifdef WIN32
	WSADATA	wsa_data ;

	/* Initialize the winsock lib ( version 2.2 ) */
	if ( WSAStartup(MAKEWORD(2,2), &wsa_data) == SOCKET_ERROR ){
		printf("WSAStartup() failed : %d\n", GetLastError()) ;
		return 1 ;
	}
#endif /* WIN32 */

netlib_init();
set_defaults();
scan_cmd_line(argc,argv);

if (debug) {
	dump_globals();
	install_signal_catchers();
}

if (debug) {
	printf("remotehost is %s and port %d\n",host_name,test_port);
	fflush(stdout);
}


establish_control(host_name,test_port);

if (0) {
}
#ifdef DO_SCTP
else if (strcasecmp(test_name,"SCTP_STREAM") == 0) {
	send_sctp_stream(host_name);
}
else if (strcasecmp(test_name,"SCTP_MAERTS") == 0) {
	send_sctp_maerts(host_name);
}
#ifdef HAVE_SENDFILE
else if (strcasecmp(test_name,"SCTP_SENDFILE") == 0) {
	sendfile_sctp_stream(host_name);
}
#endif /* HAVE_SENDFILE */
else if (strcasecmp(test_name,"SCTP_RR") == 0) {
	send_sctp_rr(host_name);
}
else if (strcasecmp(test_name,"SCTP_CRR") == 0) {
	send_sctp_conn_rr(host_name);
}
else if (strcasecmp(test_name,"SCTP_CC") == 0) {
	send_sctp_cc(host_name);
}
#ifdef DO_1644
else if (strcasecmp(test_name,"SCTP_TRR") == 0) {
	send_sctp_tran_rr(host_name);
}
#endif /* DO_1644 */
#ifdef DO_NBRR
else if (strcasecmp(test_name,"SCTP_NBRR") == 0) {
	send_sctp_nbrr(host_name);
}
#endif /* DO_NBRR */
#endif	/* DO_SCTP */
else if (strcasecmp(test_name,"TCP_STREAM") == 0) {
	send_tcp_stream(host_name);
}
else if (strcasecmp(test_name,"TCP_MAERTS") == 0) {
	send_tcp_maerts(host_name);
}
#ifdef HAVE_SENDFILE
else if (strcasecmp(test_name,"TCP_SENDFILE") == 0) {
	sendfile_tcp_stream(host_name);
}
#endif /* HAVE_SENDFILE */
else if (strcasecmp(test_name,"TCP_RR") == 0) {
	send_tcp_rr(host_name);
}
else if (strcasecmp(test_name,"TCP_CRR") == 0) {
	send_tcp_conn_rr(host_name);
}
else if (strcasecmp(test_name,"TCP_CC") == 0) {
	send_tcp_cc(host_name);
}
#ifdef DO_1644
else if (strcasecmp(test_name,"TCP_TRR") == 0) {
	send_tcp_tran_rr(host_name);
}
#endif /* DO_1644 */
#ifdef DO_NBRR
else if (strcasecmp(test_name,"TCP_NBRR") == 0) {
	send_tcp_nbrr(host_name);
}
#endif /* DO_NBRR */
else if (strcasecmp(test_name,"UDP_STREAM") == 0) {
	send_udp_stream(host_name);
}
else if (strcasecmp(test_name,"UDP_RR") == 0) {
	send_udp_rr(host_name);
}
else if (strcasecmp(test_name,"LOC_CPU") == 0) {
	loc_cpu_rate();
}
else if (strcasecmp(test_name,"REM_CPU") == 0) {
	rem_cpu_rate();
}
#ifdef DO_DLPI
else if (strcasecmp(test_name,"DLCO_RR") == 0) {
	send_dlpi_co_rr(host_name);
}
else if (strcasecmp(test_name,"DLCL_RR") == 0) {
	send_dlpi_cl_rr(host_name);
}
else if (strcasecmp(test_name,"DLCO_STREAM") == 0) {
	send_dlpi_co_stream(host_name);
}
else if (strcasecmp(test_name,"DLCL_STREAM") == 0) {
	send_dlpi_cl_stream(host_name);
}
#endif /* DO_DLPI */
#ifdef DO_UNIX
else if (strcasecmp(test_name,"STREAM_RR") == 0) {
	send_stream_rr(host_name);
}
else if (strcasecmp(test_name,"DG_RR") == 0) {
	send_dg_rr(host_name);
}
else if (strcasecmp(test_name,"STREAM_STREAM") == 0) {
	send_stream_stream(host_name);
}
else if (strcasecmp(test_name,"DG_STREAM") == 0) {
	send_dg_stream(host_name);
}
#endif /* DO_UNIX */
#ifdef DO_FORE
else if (strcasecmp(test_name,"FORE_STREAM") == 0) {
	send_fore_stream(host_name);
}
else if (strcasecmp(test_name,"FORE_RR") == 0) {
	send_fore_rr(host_name);
}
#endif /* DO_FORE */
#ifdef DO_HIPPI
else if (strcasecmp(test_name,"HIPPI_STREAM") == 0) {
	send_hippi_stream(host_name);
}
else if (strcasecmp(test_name,"HIPPI_RR") == 0) {
	send_hippi_rr(host_name);
}
#endif /* DO_HIPPI */
#ifdef DO_XTI
#ifdef DO_XTI_SCTP
else if (strcasecmp(test_name,"XTI_SCTP_STREAM") == 0) {
	send_xti_sctp_stream(host_name);
}
else if (strcasecmp(test_name,"XTI_SCTP_RR") == 0) {
	send_xti_sctp_rr(host_name);
}
#endif	/* DO_XTI_SCTP */
else if (strcasecmp(test_name,"XTI_TCP_STREAM") == 0) {
	send_xti_tcp_stream(host_name);
}
else if (strcasecmp(test_name,"XTI_TCP_RR") == 0) {
	send_xti_tcp_rr(host_name);
}
else if (strcasecmp(test_name,"XTI_UDP_STREAM") == 0) {
	send_xti_udp_stream(host_name);
}
else if (strcasecmp(test_name,"XTI_UDP_RR") == 0) {
	send_xti_udp_rr(host_name);
}
#endif /* DO_XTI */
#ifdef DO_LWP
else if (strcasecmp(test_name,"LWPSTR_STREAM") == 0) {
	send_lwpstr_stream(host_name);
}
else if (strcasecmp(test_name,"LWPSTR_RR") == 0) {
	send_lwpstr_rr(host_name);
}
else if (strcasecmp(test_name,"LWPDG_STREAM") == 0) {
	send_lwpdg_stream(host_name);
}
else if (strcasecmp(test_name,"LWPDG_RR") == 0) {
	send_lwpdg_rr(host_name);
}
#endif /* DO_LWP */
#ifdef DO_IPV6
#ifdef DO_IPV6_SCTP
else if (strcasecmp(test_name,"SCTPIPV6_STREAM") == 0) {
	send_sctpipv6_stream(host_name);
}
else if (strcasecmp(test_name,"SCTPIPV6_RR") == 0) {
	send_sctpipv6_rr(host_name);
}
else if (strcasecmp(test_name,"SCTPIPV6_CRR") == 0) {
	send_sctpipv6_rr(host_name);
}
#endif	/* DO_IPV6_SCTP */
else if (strcasecmp(test_name,"TCPIPV6_STREAM") == 0) {
	send_tcpipv6_stream(host_name);
}
else if (strcasecmp(test_name,"TCPIPV6_RR") == 0) {
	send_tcpipv6_rr(host_name);
}
else if (strcasecmp(test_name,"TCPIPV6_CRR") == 0) {
	send_tcpipv6_rr(host_name);
}
else if (strcasecmp(test_name,"UDPIPV6_STREAM") == 0) {
	send_udpipv6_stream(host_name);
}
else if (strcasecmp(test_name,"UDPIPV6_RR") == 0) {
	send_udpipv6_rr(host_name);
}
#endif /* DO_IPV6 */
#ifdef DO_DNS
else if (strcasecmp(test_name,"DNS_RR") == 0) {
	send_dns_rr(host_name);
}
#endif /* DO_DNS */
else {
	printf("The test you requested is unknown to this netperf.\n");
	printf("Please verify that you have the correct test name, \n");
	printf("and that test family has been compiled into this netperf.\n");
	exit(1);
}

shutdown_control();

#ifdef WIN32
	/* Cleanup the winsock lib */
	WSACleanup();
#endif

return(0);
}


